subsys/net/lib/lwm2m/lwm2m_obj_firmware.c

```diff
@@ -59,11 +60,85 @@ static struct lwm2m_engine_obj_inst inst;
 static struct lwm2m_engine_res_inst res[FIRMWARE_MAX_ID];
 
 static lwm2m_engine_set_data_cb_t write_cb;
+static lwm2m_engine_exec_cb_t update_cb;
 
 #ifdef CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT
 extern int lwm2m_firmware_start_transfer(char *package_uri);
 #endif
 
+u8_t lwm2m_firmware_get_update_state(void)
+{
+	return update_state;
+}
+
+void lwm2m_firmware_set_update_state(u8_t state)
+{
+	/* Check LWM2M SPEC appendix E.6.1 */
+#if defined(CONFIG_ASSERT)
+	switch (state) {
+	case STATE_DOWNLOADING:
+		assert(update_state == STATE_IDLE);
+		break;
+	case STATE_DOWNLOADED:
+		assert(update_state == STATE_DOWNLOADING ||
+		       update_state == STATE_UPDATING);
+		break;
+	case STATE_UPDATING:
+		assert(update_state == STATE_DOWNLOADED);
+		break;
+	case STATE_IDLE:
+		break;
+	default:
+		assert(0 && "Shouldn't happen");
+	}
+#endif
+	update_state = state;
+	SYS_LOG_DBG("Update state = %d", update_state);
+}
+
+u8_t lwm2m_firmware_get_update_result(void)
+{
+	return update_result;
+}
+
+void lwm2m_firmware_set_update_result(u8_t result)
+{
+	/* Check LWM2M SPEC appendix E.6.1 */
+	switch (result) {
+	case RESULT_DEFAULT:
+		lwm2m_firmware_set_update_state(STATE_IDLE);
+		break;
+	case RESULT_SUCCESS:
+		assert(update_state == STATE_UPDATING);
```

Comment:
Once again, I'm not sure an assert is the right behavior here.  Consider a strongly worded error message?